charms/istio-pilot/lib/charms/istio_pilot/v0/ingress.py

```diff
@@ -0,0 +1,268 @@
+        self,
+        charm: CharmBase,
+        relation_name: str = DEFAULT_RELATION_NAME,
+        *,
+        port: int = None,
+        service: str = None,
+        prefix: str = None,
+        rewrite: str = None,
+        namespace: str = None,
+        per_unit_routes: bool = False,
+    ):
+        """Constructor for IngressRequirer.
+
+        The request args can be used to specify the ingress properties when the
+        instance is created. If any are set, at least `port` is required, and
+        they will be sent to the ingress provider as soon as it is available.
+        All request args must be given as keyword args.
+
+        Args:
+            charm: the charm that is instantiating the library.
+            relation_name: the name of the relation endpoint to bind to
+                (defaults to "ingress"; relation must be of interface type
+                "ingress" and have "limit: 1")
+        Request Args:
+            service: the name of the target K8s service to route to; defaults to the
+                charm's automatically created service (i.e., the application name)
+            port: the port of the service (required)
+            prefix: the path used to match this service for requests to the gateway;
+                must not conflict with other services; defaults to f"/{service}/"
+            rewrite: the path on the target service to map the request to; defaults
+                to "/"
+            namespace: the namespace the service is in; default to the current model
+            per_unit_routes: whether or not to create URLs which map to specific units;
+                the URLs will have their own prefix of f"{prefix}-unit-{unit_num}" (with
+                tailing slashes handled appropriately)
+        """
+        super().__init__(charm, f"ingress-requirer-{relation_name}")
+        self.charm = charm
+        self.relation_name = relation_name
+
+        self._validate_relation_meta()
+
+        self.status = self._get_status()
+
+        self._request_args = {
+            "port": port,
+            "service": service,
+            "prefix": prefix,
+            "rewrite": rewrite,
+            "namespace": namespace,
+            "per_unit_routes": per_unit_routes,
+        }
+        if any(self._request_args.values()) and not self._request_args["port"]:
+            raise TypeError("Missing required argument: 'port'")
+
+        if charm.unit.is_leader():
+            self.framework.observe(charm.on[relation_name].relation_created, self._check_provider)
+            self.framework.observe(charm.on[relation_name].relation_changed, self._check_provider)
+            self.framework.observe(charm.on[relation_name].relation_broken, self._lost_provider)
+            self.framework.observe(charm.on.leader_elected, self._check_provider)
+
+    def _get_status(self):
+        if not self.charm.unit.is_leader():
+            return None
+        if not self.charm.model.relations[self.relation_name]:
```

Comment:
This also breaks access to per-unit routing info to non-leader units

Reply:
How?  Non-leaders can't send an ingress request, so they shouldn't need to care about the status.

Reply:
> Non-leaders can't send an ingress request, so they shouldn't need to care about the status

Non-leaders are supposed to be able to follow the events, right? (At least, Prometheus does :-) ) Ingress events are fired only in the leader, though. So I do not think this limitation is acceptable in per-unit routing.